### PR TITLE
Allow for multiple keysets configured

### PR DESCRIPTION
**Pull request by johnlanda**
# Overview

This change allows for the JWT auth backend to be configured with multiple JWKS URLs as the result of a customer request.

This change should allow for current configuration using `jwks_url` and `jwks_ca_pem` to work as normal. Users may optionally choose to instead specify sets of `jwks_pairs`. If `jwks_pairs` are specified then the underlying `Validator` from the cap library will loop over the available `jwks_pairs` in order to attempt to validate the JWT.

# Design of Change
This change updates the config for the `jwtAuthBackend` in the `path_config.go` file. It also updates the `backend.go` file to ensure that when using `jwks_pairs` we iterate over the JWKS uris and CA PEMs to initialize all key sets.

# Related Issues/Pull Requests
[ ] [Cap PR #128](https://github.com/hashicorp/cap/pull/128)

# Contributor Checklist
[ ] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
[My Docs PR Link](link)
[Example](https://github.com/hashicorp/vault/commit/2715f5cec982aabc7b7a6ae878c547f6f475bba6)
[ ] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
[x] Backwards compatible

### CODE REVIEW

File: backend.go
Comment (johnlanda) [inline backend.go:158]: Is there a pattern for nested config in the path config? Ideally I would like these to be more like a list of structs than a map of key-value pairs.

File: go.mod
Comment (johnlanda) [inline go.mod:101]: To be replaced once the Cap PR is merged.

https://github.com/hashicorp/cap/pull/128

File: path_config.go
Comment (johnlanda) [inline path_config.go:76]: Like the comment above: it would probably be a more usable config if there is a good way to do a slice of struct here instead of a map. I was having a hard time getting that to parse properly.

File: path_config.go
Comment (johnlanda) [inline path_config.go:265]: I left the `jwks_url` and `jwks_ca_pem` config in place to maintain backwards compatibility. It could be nice just to specify a single `jwks_pair` instead but I think the backwards compatibility was more important.

File: path_config.go
Comment (johnlanda) [inline path_config.go:312]: updating deprecated method

File: path_config.go
Comment (johnlanda) [inline path_config.go:413]: This is just the existing logic split into a helper function to be used for the single and multi JWKS URL cases.

File: backend.go
Comment (austingebauer) [reply backend.go:158]: Unfortunately there isn't a great pattern for complex/nested data types in the Vault API. You can use `framework.TypeMap` as you have, but it has the limitation that it's not usable by the Vault CLI without converting the entire request body to JSON. Happy to show you an example of this limitation.

I'm wondering if using `framework.TypeKVPairs` would be better. That would make the data type `map[string]string` where the key is the JWKS URL and the value the CA PEM. It also allows for using the Vault CLI more naturally.

File: path_config.go
Comment (austingebauer) [reply path_config.go:76]: Ya.. it's a challenge today with Vault backends. There are some ways to make it work but they require us to manually parse outside of the backend framework. Do we anticipate needing any configuration in addition to the JWKS URL and CA?

File: path_config.go
Comment (austingebauer) [reply path_config.go:265]: +1 on that decision

File: path_config.go
Comment (johnlanda) [reply path_config.go:76]: I don't expect any additional config to be added to the pairs.

File: backend.go
Comment (johnlanda) [reply backend.go:158]: `TypeKVPairs` sounds like an improvement over the current setup at least. I don't think the config files will read super intuitively unfortunately since the fields within each `jwks_pair` will be effectively unlabeled.

Comment (austingebauer) [reply backend.go:158]: Ya, I see what you're saying. Do you have a rough example of what that config file would look like? It might help me think of something better 🤔

Comment (johnlanda) [reply backend.go:158]: If there is anyway to make it like a repeated block in yaml where it would be something like 
```
jwks_pairs:
  - jwks_url: www.foobar.com/wellknown
    jwks_ca_pem: adfjkafklasdfs232flfdskja
  - jwks_url: www.barbaz.com/wellknown
    jwks_ca_pem: aweeqrrnrqewrkljerwqwr
```

Comment (johnlanda) [reply backend.go:158]: Looking through the nomad docs for setting up vault doesn't show setting the ca cert. Is that something that vault does automatically? If so we could just make this a list of urls

Comment (austingebauer) [reply backend.go:158]: I'm trying something out with the repeated block idea. I'll follow up here shortly!

> Looking through the nomad docs for setting up vault doesn't show setting the ca cert. Is that something that vault does automatically? If so we could just make this a list of urls

Able to share a link to which doc you're looking at? It's not required if your IdP (Nomad in this case) is not TLS protected or the system trust store where Vault is running has already trusted the CA of the IdP.

Comment (jkirschner-hashicorp) [reply backend.go:158]: > Looking through the nomad docs for setting up vault doesn't show setting the ca cert. Is that something that vault does automatically?

Even if Nomad doesn't now, or maybe doesn't in its "easy getting starts commands / docs", setting the CA cert might be something we want to support (even if some people leave it as an empty string). I defer to you both, though.

Comment (johnlanda) [reply backend.go]: Updated so that the config for `jwks_pairs` is effectively a slice of structs and the fields are named for easier reading of config.

Comment (johnlanda) [reply backend.go]: New config looks like this
```vault write auth/jwt/config -<<EOF 
{
    "jwks_pairs": [
        { "jwks_url": "foobar.com", "jwks_ca_pem": "asdfadsfasdf" },
        { "jwks_url": "barbaz.com", "jwks_ca_pem": "adfsasdfasdf" }
    ]
}
EOF
```

**Review by austingebauer — COMMENTED** [review]
Leaving a couple of small comments as I'm looking through this. Looking good!

File: jwks_pairs.go
Comment (johnlanda) [reply jwks_pairs.go:13]: I often use `<=` instead of `==` as it is more restrictive than `==`. Though `len` should always return a positive integer so I'll defer to your preference.

File: jwks_pairs.go
Comment (johnlanda) [reply jwks_pairs.go:19]: Ack. Will update and add a test.

File: backend.go
Comment (austingebauer) [reply backend.go]: Thanks for the example, @johnlanda!

File: jwks_pairs.go
Comment (johnlanda) [reply jwks_pairs.go]: Added in f7733e98b32797fde5277d2056f794e8d53da445

File: jwks_pairs.go
Comment (johnlanda) [reply jwks_pairs.go:15]: This is probably an overly defensive habit so as not to assume that the function called always returns a positive number.

Comment (austingebauer) [reply jwks_pairs.go:15]: Makes sense! I think it's okay to leave as is :)

File: backend.go
Comment (austingebauer) [inline backend.go:164]: I believe the error is shadowed here such that if errors occurred, they wouldn't be appended to the outer error. I don't think we'd end up in the `if err != nil {` check below this switch case.

File: backend.go
Comment (austingebauer) [inline backend.go:162]: Missing a `continue`? A nil `keySet` would be appended for the error case.

File: backend.go
Comment (johnlanda) [reply backend.go:164]: Good catch. Looks like we can rename that error to something like `pairsErr` and just check and return it before the loop if not nil.

File: backend.go
Comment (johnlanda) [reply backend.go:162]: 🤦

**Review by austingebauer — APPROVED** [review]
Thanks again for this contribution, @johnlanda! LGTM

**Pull request merged**
